Option to provide additional host configuration

sweiler opened the pull request:
In the Java implementation, it is possible to modify the `docker run` command before it is being executed with `withCreateContainerCmdModifier`. So far there is no similar option in testcontainers-node.

This PR adds an option to add additional host parameters, exposing the underlying dockerode API. This is very similar to how the Java implementation with the above mentioned `withCreateContainerCmdModifier` exposes the underlying docker-java API. (See [here](https://github.com/testcontainers/testcontainers-java/blob/6622578bed5e66bfb979b9bfeeb0e22321e21266/core/src/main/java/org/testcontainers/containers/GenericContainer.java#L1351))

The specific reason why we would need this feature, is that we want to run a test using testcontainers on a docker host with splunk set as the log driver. But Splunk is a write-only log driver, which leads to errors when waiting for a log line in testcontainers. With this PR, we could set the log-driver for the test explicitly to json-file.

cristianrgreco commented:
Hi @sweiler, thanks for the PR!

Would you mind also updating the README to show an example of how it can be used?

I'm not sure if it should be called `additionalHostConfig` instead of for example `hostConfig`, or even something more specific in this case like `withLogConfig`. One of the goals of testcontainers is to abstract away the details of Docker, and using `withHostConfig` would mean a user would need to look at the docs to know what (including what types) can be provided.

I think in this case providing a specific, typed API for building the container is better. What do you think?

sweiler commented:
Thanks for the quick response :)

I'm happy to update the README as well.
I decided to go with the very generic host config, because I didn't want to add an option only for my own case. Also I thought it would be more aligned with the way testcontainers-java does it, which is also the very generic 'all options available' way. But if you say you'd like a specific `withLogConfig` option more, I'm more than happy to change my PR accordingly.

cristianrgreco commented:
I think a typed `withLogConfig` is what we want, ideally users shouldn't have to read the Docker docs to know what values to provide

sweiler commented:
Hey @cristianrgreco, I just updated the PR to contain the withLogConfig. I also changed the naming to be more similar to the docker run command and the docker daemon config, instead of the Dockerode naming.
Also I added a paragraph to the README.

What do you think?

cristianrgreco commented on README.md:
````diff
@@ -247,3 +247,25 @@ const container = await new GenericContainer("redis")
   .start();
 ```
+
+## Log Config
````
This should probably be further up with the other examples, it's currently in the wait strategies section

sweiler replied:
You're right, I didn't quite get the structure I have to admit. Moved it further up to a point where I think it makes sense.

cristianrgreco commented on README.md:
```diff
@@ -247,3 +247,25 @@ const container = await new GenericContainer("redis")
+const { GenericContainer, Wait } = require("testcontainers");
+
+const container = await new GenericContainer("redis")
+  .withExposedPorts(6379)
```
The `Wait` import and `withExposedPorts` could be removed to simplify the example

cristianrgreco commented on src/docker-client.ts:
```diff
@@ -53,6 +53,15 @@ export type BindMount = {
 type DockerodeBindMount = string;
 
+export type LogConfig = {
+  logDriver: string;
```
I was going through the Docker docs to see what types we can add to this. It seems there's a limited number of logDrivers, but each one has different options. For example for Splunk, we would need to provide splunk-token and splunk-url (https://docs.docker.com/config/containers/logging/splunk/). As a result, what do you think about adding log-driver specific implementations? For example:

`withLogDriverForSplunk(token: String, url: String)`

I'm not sure what your use-case is, but you could add your specific log-driver implementation, and specify the mandatory options?

sweiler replied:
Hm, I tend to disagree, as I think that people who need this feature also know quite well what log-driver they want and how the options they need for it are called, as they probably have this configuration already somewhere else.
On the other hand, I think it is quite possible that the use-case for most people is the same I had, which is that you configured a special log-driver on your docker host, but want to use the plain default 'json-file' logger with no special options for testcontainers. So what about having a withDefaultLogDriver() which has no more options and just resets the log driver to json-file with no options, which is the default in docker?

sweiler replied:
@cristianrgreco any updates on this?

cristianrgreco commented on src/docker-client.ts:
```diff
@@ -53,6 +53,15 @@ export type BindMount = {
+};
+type DockerodeLogConfig = {
+  Type: string;
+  Config: any;
```
I think this will always be an `object`

cristianrgreco commented on src/generic-container.test.ts:
```diff
@@ -160,6 +160,30 @@ describe("GenericContainer", () => {
   });
 
+  it("should set log config", async () => {
+    const containerName = "special-test-container";
```
Can remove the name expectations, the assertion we care about is on https://github.com/testcontainers/testcontainers-node/pull/58/files#diff-9c365222db7442c7a59145d8c7b621e2R179

sweiler commented:
Any updates on this @cristianrgreco ?

cristianrgreco commented:
Sorry @sweiler, completely forgot about this. I think we're missing the `withLogConfig` method in the `TestContainer` interface. Once added I think we're good to go.

sweiler commented:
No worries, thanks for responding :) I added it now to the interface, so I would be happy to get this merged as it is now.

But I also think your point in one of the comments above about this requiring every user now to know the driver options is still valid. So If you woud prefer the variant `withDefaultLogDriver()` that I proposed over this one, I'm also happy to change it again.

cristianrgreco commented:
Thanks @sweiler. 

I think I would prefer `withDefaultLogDriver`. As it would allow more specific log driver implementations to be added in the future, like `withSplunkLogDriver()`, etc.

 Although the implementation as it is would allow someone to configure the log driver however they see fit, my concern is that testcontainers isn't helping by abstracting away the Docker internals, it's surfacing them - the user **has** to read the Docker docs to figure out what to provide here

sweiler commented:
Cool, I changed the PR now accordingly. Could you also check again if I documented it in the README in a way users understand it?

cristianrgreco commented:
Thanks @sweiler, LGTM. Are you happy for me to merge?

sweiler commented:
Cool, very happy if you'd merge it :) Thanks for the nice collaboration experience 😊

cristianrgreco commented:
Thank you @sweiler! The change is available in version 2.7.0.
